torch/_inductor/decomposition.py

```diff
@@ -705,3 +706,50 @@ def _softmax_backward_data(grad_output, output, dim, input_dtype):
     if grad_output.dtype != input_dtype:
         grad_input = grad_input.to(input_dtype)
     return grad_input.contiguous()
+
+
+@register_decomposition(aten.index_reduce)
+def index_reduce(
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    if reduction_type == "mean" and not needs_fallback_due_to_atomic_add_limitations(
+        self.dtype
+    ):
+        tmp = (torch.ones_like if include_self else torch.zeros_like)(self)
+        counts = tmp.index_add(dim, index, torch.ones_like(src))
+        out = self if include_self else tmp  # tmp is torch.zeros_like(self)
+        out = out.index_add(dim, index, src)
```

Comment:
I think the accuracy failures are because we need to zero out the indexed locations before calling `index_add` in the the `include_self=False` case:
https://github.com/pytorch/pytorch/blob/5a15cbfa440e87f8439d1f28b3d8df0c51daf088/aten/src/ATen/native/cuda/Indexing.cu#L1026-L1050

Reply:
Hmm, in the `include_self=False` case, `out` is defined as `torch.zeros_like(self)` (see the definition of `tmp`). IIUC, `out` is already initialized with zeros. So, I don't see a problem here unless calling `tmp.index_add` has side-effects to `tmp` data.

Reply:
There appears to be a bug that leads to these failures. For instance, for inputs with `shape=(1,)` and having `counts=[1]`, both `mask = counts == 0` and `mask = counts == 1` lead to the same result: `mask == [True]`.

Reply:
I don't think out should be initialized with zeros when `include_self = False`, I can't see that in the eager code.

Reply:
Also how did you inspect the mask? I wouldn't expect that to be a real buffer in the generated code.

Reply:
> I don't think out should be initialized with zeros when include_self = False, I can't see that in the eager code.

`out` is initialized with zeros in
https://github.com/pytorch/pytorch/blob/5f378e1853558d603028c2062373869c6f192885/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L1016

Reply:
Anyway, using
```
mask = counts < 1
```
fixed the issue (tested locally).

Reply:
Yes precisely, `index_fill` not `fill_` so it's not the whole tensor that's zeroed, only the indexes specified.

Reply:
That is a very good point, thanks! When using `index_fill`, the `index_reduce` implementation will simplify.